### ISSUE TITLE
Use of restricted API detected

### ISSUE THREAD

**Issue by spicyseaweed** [nus-cs2030/2122-s2#83]
# Summary

Hi everyone, CodeCrunch failed my submission cuz I imported java.util.Collection
But I thought we need the collection class for values() method??
Can anyone help me with this issue?

## Description:

Clarifications or details with respect to the issue you are facing

## Code Snippets

Attach code snippets (if any)

## Screenshots

![image](https://user-images.githubusercontent.com/94048668/157865000-d0d35578-54e7-4351-bd6a-af197f204802.png)

**Comment by Yang990121** [comment]
I think you level 5, you loop through the collection instead of looping through the dictionary?

**Comment by Yang990121** [comment]
or you can try to submit code crunch level by level to see which level is causing the issue

**Comment by huangrz0702** [comment]
U can send ur immutablelist class first and see whether there is error

**Comment by spicyseaweed** [comment]
Thanks alot for the reply!

**Comment by huangrz0702** [comment]
if not it might be ur level 4/5 problem?

**Comment by spicyseaweed** [comment]
Thank you Yang990121! Thank you huangrz0702!

**Comment by spicyseaweed** [comment]
Thanks I passed !!!!!!

**Comment by shotnothing** [comment]
I had the exact same issue but fixed it after replacing this method:

![image](https://user-images.githubusercontent.com/16388604/157872572-e5913ca9-26c5-4081-8024-4912274ec3b0.png)

with one that does not depend on map or list. I suspect its to do with the enhanced for-loop indirectly calling a collection method but I have no idea why.

**Comment by tygq13** [comment]
Hi, I also have this error when I use java.util.Set. So we are not supposed to use either “Set" or "Collection" or "ArrayList" to solve this lab?

**Comment by shotnothing** [comment]
> Hi, I also have this error when I use java.util.Set. So we are not supposed to use either “Set" or "Collection" or "ArrayList" to solve this lab? If so what's supposed to be the return type for keySet(), values() and entrySet()?

List, List.copyOf(whatever) works

**Comment by spicyseaweed** [comment]
> 

hi, i think we cannot use ArrayList! but we need set and collection! I realized the problem was not because of my import java. util.collection, but the problem is I shouldn't use it in my level5(module class toString())

**Comment by spicyseaweed** [comment]
> 

hi i think we cant use arraylist anymore, there's a ImList class for us to use it in place of arraylist.

**Comment by lukasleeyo** [comment]
Hi, just to add that I also encountered this issue. I tried using the code below:
```String result = String.format("%s: %s", this.title, this.assessments.values());
   return result.replace('[', '{').replace(']', '}');```

But I think it will trigger this issue because when I use ```.values()``` method, it is indirectly using the ```java.util.Collection```.

So like @shotnothing said, you need to loop through the iterator of the ImmutableMap and get each Assessment value individually.

**Comment by szeying02** [comment]
In general, I believe that we aren't allowed to call any methods from the Set and Collection object / class (even toString()), but we are allowed to return it.

**Comment by YeoShenKai** [comment]
> In general, I believe that we aren't allowed to call any methods from the Set and Collection object / class (even toString()), but we are allowed to return it.

Yes, I believe this is the case. The isPresent() and get() methods for Optional cannot be used. This means that you cannot create an Optional class until the return statement.

**Comment by shotnothing** [comment]
> Hi, just to add that I also encountered this issue. I tried using the code below: `String result = String.format("%s: %s", this.title, this.assessments.values());` `return result.replace('[', '{').replace(']', '}');`
> 
> But I think it will trigger this issue because when I use `.values()` method, it is indirectly using the `java.util.Collection`.
> 
> So like @shotnothing said, you need to loop through the iterator of the ImmutableMap and get each Assessment value individually.

Ohh thats weird, because my final Module.java had 
`String out = map.values().toString();`
as part of `toString()`, and it got thru codecrunch.
